src/python/pants/base/hash_utils.py

```diff
@@ -39,6 +48,37 @@ def hash_file(path, digest=None):
   return digest.hexdigest() if PY3 else digest.hexdigest().decode('utf-8')
 
 
+class CoercingEncoder(json.JSONEncoder):
+  """An encoder which performs coercions in order to serialize many otherwise illegal objects.
+
+  The python documentation (https://docs.python.org/2/library/json.html#json.dumps) states that
+  dict keys are coerced to strings in json.dumps, but this appears to be incorrect -- it throws a
+  TypeError on things we might to throw at it, like a set, or a dict with tuple keys.
+  """
+
+  def _maybe_encode(self, key_obj):
+    """This isn't strictly necessary, but this is the behavior I might expect."""
```

Comment:
Who is "I"? This is a super strange doc comment _and_ it's on a private method. Also - why add this complexity? You've now dipped your toes into bytes vs str vs who knows what the underlying encoder does - and that's the point - let it do its job as much as possible - it should know what its doing.

Reply:
I'll have you know I had this exact thought writing it, but didn't really end up checking myself. I believe I did this because I was seeing infinite recursion and the "behavior" I was referring to was not recursing when the input is a string. I didn't then and I don't know now whether the recursion is indicative of an error I made in the implementation, or an expected thing to check. I will try reverting this and seeing if the infinite recursion truly must occur.

Reply:
This has been renamed to `_maybe_encode_dict_key()`, with a comment explaining what it does a little more coherently.